File: labs/lab13/Min_Cost_to_Connect_All_Points.py
import heapq

class Solution:
    def minCostConnectPoints(self, points):
        n = len(points)
        # The min-heap (priority queue)
        min_heap = []
        # Keep track of the visited points
        visited = [False] * n
        # Start from the first point
        visited[0] = True
        total_cost = 0
        edges_used = 0
        
        # Function to calculate the Manhattan distance between two points
        def manhattan_distance(p1, p2):
            return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])
        
        # Add the edges from point 0 to the heap
        for i in range(1, n):
            cost = manhattan_distance(points[0], points[i])
            heapq.heappush(min_heap, (cost, i))
        
        # Prim's Algorithm to find MST
        while edges_used < n - 1:
            cost, point = heapq.heappop(min_heap)
            
            # If the point is already visited, we skip it
            if visited[point]:
                continue
            
            # Otherwise, we add the cost and mark the point as visited
            visited[point] = True
            total_cost += cost
            edges_used += 1
            
            # Add new edges from the current point to the heap
            for i in range(n):
                if not visited[i]:
                    new_cost = manhattan_distance(points[point], points[i])
                    heapq.heappush(min_heap, (new_cost, i))
        
        return total_cost

# Example usage
solution = Solution()
print(solution.minCostConnectPoints([[0,0],[2,2],[3,10],[5,2],[7,0]]))  # Output: 20
print(solution.minCostConnectPoints([[3,12],[-2,5],[-4,1]]))  # Output: 18

File: labs/lab13/Min_Cost_to_Connect_All_Points.js
class MinHeap {
  constructor() {
    this.heap = [];
  }

  // Push an element to the heap
  add([cost, point]) {
    this.heap.push([cost, point]);
    this.heapifyUp();
  }

  // Remove and return the minimum element (root of the heap)
  remove() {
    if (this.heap.length === 1) {
      return this.heap.pop();
    }

    const min = this.heap[0];
    this.heap[0] = this.heap.pop();
    this.heapifyDown();
    return min;
  }

  // Restore the heap property from bottom to top
  heapifyUp() {
    let index = this.heap.length - 1;
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2);
      if (this.heap[parentIndex][0] <= this.heap[index][0]) break;
      [this.heap[parentIndex], this.heap[index]] = [
        this.heap[index],
        this.heap[parentIndex],
      ];
      index = parentIndex;
    }
  }

  // Restore the heap property from top to bottom
  heapifyDown() {
    let index = 0;
    const length = this.heap.length;
    while (index < length) {
      let leftChildIndex = 2 * index + 1;
      let rightChildIndex = 2 * index + 2;
      let smallest = index;

      if (
        leftChildIndex < length &&
        this.heap[leftChildIndex][0] < this.heap[smallest][0]
      ) {
        smallest = leftChildIndex;
      }
      if (
        rightChildIndex < length &&
        this.heap[rightChildIndex][0] < this.heap[smallest][0]
      ) {
        smallest = rightChildIndex;
      }
      if (smallest === index) break;
      [this.heap[index], this.heap[smallest]] = [
        this.heap[smallest],
        this.heap[index],
      ];
      index = smallest;
    }
  }
}

/**
 * @param {number[][]} points
 * @return {number}
 */
var minCostConnectPoints = function (points) {
  const n = points.length;
  const visited = Array(n).fill(false); // Array to keep track of visited points
  const minHeap = new MinHeap(); // Min-heap to store the edges (cost, point index)
  let totalCost = 0;
  let edgesUsed = 0;

  // Function to calculate the Manhattan distance between two points
  const manhattanDistance = (p1, p2) => {
    return Math.abs(p1[0] - p2[0]) + Math.abs(p1[1] - p2[1]);
  };

  // Start from the first point (index 0)
  visited[0] = true;

  // Add edges from point 0 to the heap
  for (let i = 1; i < n; i++) {
    const cost = manhattanDistance(points[0], points[i]);
    minHeap.add([cost, i]);
  }

  // Prim's Algorithm to find MST
  while (edgesUsed < n - 1) {
    // Get the edge with the minimum cost
    const [cost, point] = minHeap.remove();

    // Skip if the point is already visited
    if (visited[point]) {
      continue;
    }

    // Add the cost of this edge to the total cost and mark the point as visited
    visited[point] = true;
    totalCost += cost;
    edgesUsed++;

    // Add new edges from the current point to the heap
    for (let i = 0; i < n; i++) {
      if (!visited[i]) {
        const newCost = manhattanDistance(points[point], points[i]);
        minHeap.add([newCost, i]);
      }
    }
  }

  return totalCost;
};

// Example usage
console.log(
  minCostConnectPoints([
    [0, 0],
    [2, 2],
    [3, 10],
    [5, 2],
    [7, 0],
  ])
); // Output: 20
console.log(
  minCostConnectPoints([
    [3, 12],
    [-2, 5],
    [-4, 1],
  ])
); // Output: 18
